### DISCUSSION TITLE
Difference between constraint solver's LNS with CP-SAT's LNS?

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#2829] [CP-SAT questions] [answered]
Hi experts,

Could you please share with me the diffference between constraint solver's LNS with CP-SAT's LNS subsolver?

Thanks,
John

**Comment by lperron** [comment, marked as answer]
None fundamentally. Except that the CP original version is a framework for writing lns operators, while the CP-SAT lns is a closed set of operators that we have written and that cannot be extended by users of the library.

**Reply by C8504** [reply]
Thanks Lperron, in CP original version, I saw a .cc sample simple_ls_sample.cc, does this framework support python as well?
thanks again.

**Comment by lperron** [comment]
Yes. But you should not use the original CP solver. Its performance is way
worse than CP-SAT.

**Reply by C8504** [reply]
ok, thanks.
Do we have any plan to open CP-SAT as a framework to add customized lns operators as well then?

**Comment by lperron** [comment]
Not for the time being.

**Reply by C8504** [reply]
Got it, thanks a lot!
